crates/chain/src/keychain/txout_index.rs

```diff
@@ -50,9 +51,17 @@ impl<K: Ord> Append for ChangeSet<K> {
     /// For each `last_revealed` in the given [`ChangeSet`]:
     /// If the keychain already exists, increase the index when the other's index > self's index.
     fn append(&mut self, other: Self) {
-        // We use `extend` instead of `BTreeMap::append` due to performance issues with `append`.
-        // Refer to https://github.com/rust-lang/rust/issues/34666#issuecomment-675658420
-        self.keychains_added.extend(other.keychains_added);
+        for (new_keychain, new_descriptor) in other.keychains_added {
+            if !self.keychains_added.contains_key(&new_keychain)
+                // FIXME: very inefficient
+                && self
+                    .keychains_added
+                    .values()
+                    .all(|descriptor| descriptor != &new_descriptor)
+            {
```

Comment:
🔧 Would this work as a more efficient single pass, short-circuiting way to do this?
```suggestion
            if self.keychains_added.iter().all(|(keychain, descriptor)| {
                keychain != &new_keychain && descriptor != &new_descriptor
            }) {
```

Reply:
No this code looks wrong. It's meant to check that no descriptor already in the changeset is the same as the one we're adding. The correct way to do this is a bijective map struct i.e. a BTreeMap one way and HashMap the other way. This might be handy to have in general but I guess it belongs in another PR.

Reply:
My suggested code does check that neither the new `keychain` or `descriptor` being added exists in the set of `keychains_added` just as the original code, but should be O(n) instead of O(2n). 

But I'm ok leaving this as is for now and worrying about the efficiency later if it ever becomes a problem.

Reply:
Sorry I understand what you are saying now. `contains_key` is O(log(n)) so keeping the fast one fast and separate seemed sensible.